smokedduck/operators.py

```diff
@@ -0,0 +1,356 @@
+            return self.source + join_sink_and_finalize
+        else:
+            return "JOIN " + self.source + " ON " + self.parent_join_cond + " = " + self.source + ".out_index" + join_sink_and_finalize
+
+    def get_child_join_conds(self):
+        return [self.single_op_table_name + ".in_index"]
+
+    def get_out_index(self):
+        return self.source + ".out_index"
+
+
+class BlockwiseNLJoin(Op):
+    def __init__(self, query_id, op_id, parent_join_cond):
+        super().__init__(query_id, self.get_name(), op_id, parent_join_cond)
+        self.inner_op_table_name = self.single_op_table_name + "_0"
+
+    def get_name(self):
+        return "BLOCKWISE_NL_JOIN"
+
+    def get_from_string(self):
+        if self.is_root:
+            return self.inner_op_table_name + " AS " + self.single_op_table_name
+        else:
+            return "JOIN " + self.inner_op_table_name + " AS " + self.single_op_table_name \
+                + " ON " + self.parent_join_cond + " = " + self.single_op_table_name + ".out_index"
+
+    def get_child_join_conds(self):
+        return [self.single_op_table_name + ".lhs_index", self.single_op_table_name + ".rhs_index"]
+
+    def get_out_index(self):
+        return self.single_op_table_name + ".out_index"
+
+
+class PiecewiseMergeJoin(Op):
+    def __init__(self, query_id, op_id, parent_join_cond):
+        super().__init__(query_id, self.get_name(), op_id, parent_join_cond)
+        self.inner_op_table_name = self.single_op_table_name + "_0"
+
+    def get_name(self):
+        return "PIECEWISE_MERGE_JOIN"
+
+    def get_from_string(self):
+        if self.is_root:
+            return self.inner_op_table_name + " AS " + self.single_op_table_name
+        else:
+            return "JOIN " + self.inner_op_table_name + " AS " + self.single_op_table_name \
+                + " ON " + self.parent_join_cond + " = " + self.single_op_table_name + ".out_index"
+
+    def get_child_join_conds(self):
+        return [self.single_op_table_name + ".lhs_index", self.single_op_table_name + ".rhs_index"]
+
+    def get_out_index(self):
+        return self.single_op_table_name + ".out_index"
+
+
+class HashJoin(Op):
+    def __init__(self, query_id, op_id, parent_join_cond):
+        super().__init__(query_id, self.get_name(), op_id, parent_join_cond)
+        self.probe = self.single_op_table_name + "_0"
+        self.build = self.single_op_table_name + "_1"
+        self.combine = self.single_op_table_name + "_2"  # TODO: integrate combine
+        self.finalize = self.single_op_table_name + "_3"  # TODO: integrate finalize
+
+        self.probe_name = self.single_op_table_name + "_probe"
+        self.build_name = self.single_op_table_name + "_build"
+
+    def get_name(self):
+        return "HASH_JOIN"
+
+    def get_from_string(self):
+        join_build = " JOIN " + self.build + " AS " + self.build_name + " ON " + self.probe_name + ".rhs_index = " + \
+                     self.build_name + ".out_index"
+
```

Comment:
limitation: doesn't handle left/right/outer join variants, 
e.g. select * from t1 right join t2 on t1.b=t2.b
output: 
      a    b b_2  c
0  44.0    a   a  4
1  44.0    a   a  5
2  45.0    b   b  7
3  42.0    a   a  4
4  42.0    a   a  5
5  43.0    b   b  7
6   NaN  NaN   c  6
current lineage:
   t2  t1  out_index
0   0   2          0
1   1   2          1
2   3   3          2
3   0   0          3
4   1   0          4
5   3   1          5

missing: 6   2  NaN          6

possible fix:  use left join to return tuples in probe relation that doesn't have a match with the build relation ```
join_build = " LEFT JOIN " + self.build + " AS " + self.build_name + " ON " + self.probe_name + ".rhs_index = " + \
                     self.build_name + ".out_index"

```

Reply:
It should be safe to set this to FULL OUTER JOIN for each join variant. I think this is what we need in order to support outer join variants where both the build and probe have values without any matches.

I'm not totally sure what the performance impacts are of changing this from an (implicit) INNER JOIN to a FULL OUTER JOIN in cases where there are only matches (so could be an INNER). My guess is the INNER code path is more optimized since they're so common. Would require some experimentation to find out, which I don't think is worth it for now.